File: tests/python/test_scene.py
"""
Test the Scene, Node, Animation API from Python.
"""
import gc
import weakref

import pytest
loom2d_native = pytest.importorskip("loom2d_native")
from loom2d_native import Node, Scene, Animation, Rect, Vec2


class TestNode:
    def test_default_position(self):
        n = Node()
        assert n.position == Vec2(0, 0)

    def test_set_position(self):
        n = Node()
        n.position = Vec2(100, 200)
        assert n.position == Vec2(100, 200)

    def test_xy_properties(self):
        n = Node()
        n.x = 42.0
        n.y = 99.0
        assert n.x == pytest.approx(42.0)
        assert n.y == pytest.approx(99.0)

    def test_rotation(self):
        n = Node()
        n.rotation = 1.5
        assert n.rotation == pytest.approx(1.5)

    def test_scale(self):
        n = Node()
        n.scale = Vec2(2, 3)
        assert n.scale == Vec2(2, 3)

    def test_name(self):
        n = Node("hero")
        assert n.name == "hero"

    def test_visible_default(self):
        n = Node()
        assert n.visible is True

    def test_add_child(self):
        parent = Node("parent")
        child  = Node("child")
        parent.add_child(child)
        assert len(parent.children()) == 1

    def test_world_position_with_parent(self):
        parent = Node()
        child  = Node()
        parent.position = Vec2(100, 0)
        child.position  = Vec2(50,  0)
        parent.add_child(child)
        wp = child.world_position()
        assert wp.x == pytest.approx(150.0, abs=0.01)

    def test_remove_from_parent(self):
        parent = Node()
        child  = Node()
        parent.add_child(child)
        child.remove_from_parent()
        assert len(parent.children()) == 0


class TestScene:
    def test_empty_scene(self):
        s = Scene()
        assert len(s.root().children()) == 0

    def test_add_node(self):
        s = Scene()
        s.add(Node("hero"))
        assert len(s.root().children()) == 1

    def test_clear(self):
        s = Scene()
        s.add(Node())
        s.add(Node())
        s.clear()
        assert len(s.root().children()) == 0

    def test_update_no_crash(self):
        s = Scene()
        s.add(Node())
        s.update(0.016)


class TestSubclassLifetime:
    """A Python Node subclass owned only by C++ must keep working.

    `scene.add(Enemy())` retains no Python reference, so the Python half of the
    object has to survive on the strength of the C++ shared_ptr alone. When it
    doesn't, PYBIND11_OVERRIDE silently falls back to the C++ base and the node
    quietly stops updating — no exception, no crash, just a dead entity. Every
    test below adds the node inline, the way game code naturally would.
    """

    def test_override_fires_when_python_keeps_no_reference(self):
        ticks = []

        class Ticker(Node):
            def update(self, dt):
                ticks.append(dt)

        s = Scene()
        s.add(Ticker())          # no Python reference retained
        gc.collect()             # force the issue

        s.update(0.016)
        s.update(0.016)

        assert ticks == [pytest.approx(0.016), pytest.approx(0.016)]

    def test_python_state_survives_c_plus_plus_ownership(self):
        class Counter(Node):
            def __init__(self):
                super().__init__("counter")
                self.count = 0

            def update(self, dt):
                self.count += 1

        s = Scene()
        s.add(Counter())
        gc.collect()

        for _ in range(3):
            s.update(0.016)

        # Reaching back through C++ must yield the same Python object, with its
        # subclass type and its attributes intact.
        node = s.root().children()[0]
        assert isinstance(node, Counter)
        assert node.count == 3

    def test_removed_node_is_freed(self):
        """Keeping the subclass alive must not mean leaking it forever.

        A scene that spawns and despawns entities would grow without bound if
        the fix were 'hold a reference until the scene dies'.
        """
        class Bullet(Node):
            pass

        s = Scene()
        bullet = Bullet()
        ref = weakref.ref(bullet)
        s.add(bullet)
        del bullet
        gc.collect()
        assert ref() is not None, "scene should still own the node"

        s.clear()
        gc.collect()
        assert ref() is None, "clearing the scene should free the node"


class TestAnimation:
    def test_construction(self):
        a = Animation("run")
        assert a.name == "run"
        assert a.frame_count() == 0

    def test_add_frame(self):
        a = Animation("idle")
        a.add_frame(Rect(0, 0, 32, 32), 0.1)
        assert a.frame_count() == 1

    def test_loop_default_true(self):
        a = Animation("run")
        assert a.loop is True

    def test_no_loop(self):
        a = Animation("death", loop=False)
        assert a.loop is False

    def test_update_advances_frame(self):
        a = Animation("run")
        a.add_frame(Rect(0,  0, 32, 32), 0.1)
        a.add_frame(Rect(32, 0, 32, 32), 0.1)
        a.update(0.15)
        assert a.frame_index() == 1

    def test_reset(self):
        a = Animation("run", loop=False)
        a.add_frame(Rect(), 0.1)
        a.add_frame(Rect(), 0.1)
        a.update(1.0)
        assert a.finished()
        a.reset()
        assert not a.finished()
        assert a.frame_index() == 0

    def test_add_strip(self):
        a = Animation("walk")
        a.add_strip(128, 32, 32, 32, 4)
        assert a.frame_count() == 4

File: tests/cpp/test_scene.cpp
#include <gtest/gtest.h>
#include "scene/node.hpp"
#include "scene/scene.hpp"
#include <memory>
#include <cmath>

using namespace loom;

// ── Node transform ────────────────────────────────────────────────────────────

TEST(Node, DefaultTransform) {
    Node n;
    EXPECT_EQ(n.position(), Vec2(0.f, 0.f));
    EXPECT_FLOAT_EQ(n.rotation(), 0.f);
    EXPECT_EQ(n.scale(), Vec2(1.f, 1.f));
    EXPECT_EQ(n.world_position(), Vec2(0.f, 0.f));
}

TEST(Node, SetPosition) {
    auto n = std::make_shared<Node>();
    n->set_position({100.f, 200.f});
    EXPECT_EQ(n->position(), Vec2(100.f, 200.f));
    EXPECT_EQ(n->world_position(), Vec2(100.f, 200.f));
}

TEST(Node, SetXY) {
    Node n;
    n.set_x(42.f); n.set_y(99.f);
    EXPECT_FLOAT_EQ(n.x(), 42.f);
    EXPECT_FLOAT_EQ(n.y(), 99.f);
}

TEST(Node, SetRotation) {
    Node n;
    n.set_rotation(1.5f);
    EXPECT_FLOAT_EQ(n.rotation(), 1.5f);
}

TEST(Node, SetScale) {
    Node n;
    n.set_scale(Vec2{2.f, 3.f});
    EXPECT_EQ(n.scale(), Vec2(2.f, 3.f));
}

TEST(Node, UniformScale) {
    Node n;
    n.set_scale(5.f);
    EXPECT_EQ(n.scale(), Vec2(5.f, 5.f));
}

// ── Parent-child hierarchy ────────────────────────────────────────────────────

TEST(Node, ChildWorldPosition_NoRotation) {
    auto parent = std::make_shared<Node>("parent");
    auto child  = std::make_shared<Node>("child");

    parent->set_position({100.f, 0.f});
    child->set_position({50.f, 0.f});
    parent->add_child(child);

    Vec2 wp = child->world_position();
    EXPECT_NEAR(wp.x, 150.f, 1e-4f);
    EXPECT_NEAR(wp.y,   0.f, 1e-4f);
}

TEST(Node, ChildWorldPosition_WithParentScale) {
    auto parent = std::make_shared<Node>();
    auto child  = std::make_shared<Node>();

    parent->set_position({0.f, 0.f});
    parent->set_scale(2.f);
    child->set_position({10.f, 0.f});
    parent->add_child(child);

    Vec2 wp = child->world_position();
    EXPECT_NEAR(wp.x, 20.f, 1e-4f); // 10 * scale(2)
}

TEST(Node, ChildWorldRotation_Accumulated) {
    auto parent = std::make_shared<Node>();
    auto child  = std::make_shared<Node>();

    parent->set_rotation(1.f);
    child->set_rotation(0.5f);
    parent->add_child(child);

    EXPECT_NEAR(child->world_rotation(), 1.5f, 1e-6f);
}

TEST(Node, ParentChildLink) {
    auto parent = std::make_shared<Node>("p");
    auto child  = std::make_shared<Node>("c");
    parent->add_child(child);

    EXPECT_EQ(parent->children().size(), 1u);
    EXPECT_EQ(child->parent(), parent.get());
}

TEST(Node, RemoveChild) {
    auto parent = std::make_shared<Node>();
    auto child  = std::make_shared<Node>();
    parent->add_child(child);
    EXPECT_EQ(parent->children().size(), 1u);

    parent->remove_child(child.get());
    EXPECT_EQ(parent->children().size(), 0u);
    EXPECT_EQ(child->parent(), nullptr);
}

TEST(Node, RemoveFromParent) {
    auto parent = std::make_shared<Node>();
    auto child  = std::make_shared<Node>();
    parent->add_child(child);
    child->remove_from_parent();
    EXPECT_EQ(parent->children().size(), 0u);
    EXPECT_EQ(child->parent(), nullptr);
}

TEST(Node, MultipleChildren) {
    auto parent = std::make_shared<Node>();
    for (int i = 0; i < 5; ++i)
        parent->add_child(std::make_shared<Node>());
    EXPECT_EQ(parent->children().size(), 5u);
}

TEST(Node, ReparentMovesChild) {
    auto p1 = std::make_shared<Node>();
    auto p2 = std::make_shared<Node>();
    auto child = std::make_shared<Node>();

    p1->add_child(child);
    EXPECT_EQ(p1->children().size(), 1u);

    // Adding to p2 should remove from p1
    p2->add_child(child);
    EXPECT_EQ(p1->children().size(), 0u);
    EXPECT_EQ(p2->children().size(), 1u);
    EXPECT_EQ(child->parent(), p2.get());
}

TEST(Node, VisibilityDefault) {
    Node n;
    EXPECT_TRUE(n.visible);
}

TEST(Node, NameDefault) {
    Node n;
    EXPECT_TRUE(n.name.empty());

    Node named("hero");
    EXPECT_EQ(named.name, "hero");
}

TEST(Node, DeepHierarchyWorldPosition) {
    auto root  = std::make_shared<Node>();
    auto mid   = std::make_shared<Node>();
    auto leaf  = std::make_shared<Node>();

    root->set_position({10.f, 0.f});
    mid->set_position({10.f, 0.f});
    leaf->set_position({10.f, 0.f});

    root->add_child(mid);
    mid->add_child(leaf);

    Vec2 wp = leaf->world_position();
    EXPECT_NEAR(wp.x, 30.f, 1e-4f);
}

// ── Scene ─────────────────────────────────────────────────────────────────────

TEST(Scene, AddAndClear) {
    Scene scene;
    scene.add(std::make_shared<Node>());
    scene.add(std::make_shared<Node>());
    EXPECT_EQ(scene.root().children().size(), 2u);

    scene.clear();
    EXPECT_EQ(scene.root().children().size(), 0u);
}

TEST(Scene, UpdatePropagates) {
    // Just verify it doesn't crash
    Scene scene;
    auto node = std::make_shared<Node>();
    scene.add(node);
    EXPECT_NO_THROW(scene.update(0.016f));
}
